pulsar-jms/src/main/java/com/datastax/oss/pulsar/jms/MessagePriorityGrowableArrayBlockingQueue.java

```diff
@@ -15,113 +15,38 @@
 
   static int getPriority(Message m) {
     return PulsarMessage.readJMSPriority(m, PulsarMessage.DEFAULT_PRIORITY);
   }
 
-  // for drainTo only, needs .add() method and that's it
-  private static class ForwardingCollection implements Collection<Pair<Integer, Message>> {
-
-    private final Collection<? super Message> c;
-
-    public ForwardingCollection(Collection<? super Message> c) {
-      this.c = c;
-    }
-
-    @Override
-    public int size() {
-      throw new IllegalStateException("Not supported");
-    }
-
-    @Override
-    public boolean isEmpty() {
-      throw new IllegalStateException("Not supported");
-    }
-
-    @Override
-    public boolean contains(Object o) {
-      throw new IllegalStateException("Not supported");
-    }
-
-    @Override
-    public Iterator<Pair<Integer, Message>> iterator() {
-      throw new IllegalStateException("Not supported");
-    }
-
-    @Override
-    public Object[] toArray() {
-      throw new IllegalStateException("Not supported");
-    }
-
-    @Override
-    public <T> T[] toArray(T[] a) {
-      throw new IllegalStateException("Not supported");
-    }
-
-    @Override
-    public boolean add(Pair<Integer, Message> message) {
-      return c.add(message.getRight());
-    }
-
-    @Override
-    public boolean remove(Object o) {
-      throw new IllegalStateException("Not supported");
-    }
-
-    @Override
-    public boolean containsAll(Collection<?> c) {
-      throw new IllegalStateException("Not supported");
-    }
-
-    @Override
-    public boolean addAll(Collection<? extends Pair<Integer, Message>> c) {
-      throw new IllegalStateException("Not supported");
-    }
-
-    @Override
-    public boolean removeAll(Collection<?> c) {
-      throw new IllegalStateException("Not supported");
-    }
-
-    @Override
-    public boolean retainAll(Collection<?> c) {
-      throw new IllegalStateException("Not supported");
-    }
-
-    @Override
-    public void clear() {}
-
-    @Override
-    public boolean equals(Object o) {
-      // shouldn't be used, but we need to keep spotbugs happy
-      return this == o || c.equals(o);
-    }
-
-    @Override
-    public int hashCode() {
-      return c.hashCode();
-    }
-  }
-
   private final PriorityBlockingQueue<Pair<Integer, Message>> queue;
   private final AtomicBoolean terminated = new AtomicBoolean(false);
 
+  private volatile Consumer<Message> itemAfterTerminatedHandler;
+  private final AtomicInteger[] numberMessagesByPrority = new AtomicInteger[10];
+
   private static final Comparator<Pair<Integer, Message>> comparator =
       (o1, o2) -> {
-        int priority1 = getPriority(o1.getRight());
-        int priority2 = getPriority(o2.getRight());
+        int priority1 = o1.getLeft();
+        int priority2 = o2.getLeft();
+        if (priority1 == priority2) {
+          // if priorities are equal, we want to sort by messageId
```

Comment:
Pair's compareTo does this already:
https://github.com/apache/commons-lang/blob/f4a616f50d3fd5af06981d6010b166f7e3362a87/src/main/java/org/apache/commons/lang3/tuple/Pair.java#L155-L159

Reply:
interesting, but I thing that this code performs too much work for a "compare" 

```
new CompareToBuilder().append(getLeft(), other.getLeft())
              .append(getRight(), other.getRight()).toComparison();
```

apart from that, I have to re-check, because I think that we want to order by priority DESC and by position ASC